### ISSUE TITLE
Refactor Java SDK to Cut Down on Cold Start Time

### ISSUE THREAD

**Issue by TimGustafson** [aws/aws-sdk-java-v2#2813]
### Describe the feature

TL;DR: Break the AWS Java SDK uber-classes (like DynamoDbClient) up into function-specific classes (like DynamoDbGetItem) so that the Maven Shade plugin's minimizeJar feature is effective at reducing JAR size and cold-start time.

Longer version:

The Java SDK has classes (like DynamoDbClient) that implement all the functionality of DynamoDB - things like creating tables and changing provisioned capacity, as well as PutItem, GetItem, and so on.  Most applications do not need to create tables or adjust provisioned capacity as part of their normal operation, so including that code in (let's say) a Lambda function that just queries data from a DynamoDB table adds to the high cold start penalty.

My idea is to split the DynamoDbClient class separate classes for each request.  For example, one DynamoDbGetItem, one DynamoDbPutItem, one DynamoDbCreateTable, and so on.  Then, when using Maven's Shade feature to include dependencies, one can enable the minimizeJar setting to remove unused classes, thereby removing a large chunk of DynamoDB business logic that is not relevant to my application.

I experimented with this by creating my own DynamoDbGetItem class, instrumented it with Aws4Signer and some other AWS-provided helper classes, and found that my cold-start times went from 8-10 seconds, down to a few hundred milliseconds instead.  Additionally, the JAR size for my function went from ~ 25MB down to a few hundred kilobytes.

Implementing all these classes by hand would be tedious, and it occurs to me that AWS is probably using some sort of code generator for the Java SDK, so it stands to reason that creating these extra function-specific classes would be much easier if it were done at code generation time, instead of after the fact.

### Is your Feature Request related to a problem?

The problem with using Java for Lambda is the cold-start times.  Including hundreds of extra methods and classes in the JAR file in order to take advantage of a handful of required AWS SDK functions is inefficient, and leads to longer cold start times.  Current cold start time reduction methods (such as using HttpURLConnection and getting the region from environment variables) help, but there's a lot of room for improvement still.  Only deploying the code that is needed for a given function will lead to much more efficient Java Lambdas.

### Proposed Solution

Break the AWS Java SDK uber-classes (like DynamoDbClient) up into function-specific classes (like DynamoDbGetItem) so that the Maven Shade plugin's minimizeJar feature is effective at reducing JAR size and cold-start time.

### Describe alternatives you've considered

I've considered building these class-per-function clients myself, but realized the task can be done more easily by the code generator that builds the AWS Java SDK more efficiently.

### Acknowledge

- [X] I may be able to implement this feature request

### AWS Java SDK version used

2.17.x

### JDK version used

Coretto 11 on Lambda

### Operating System and version

Lambda

**Comment by debora-ito** [comment]
Hi @TimGustafson thank you for reaching out. Apologies for the delay in reply.

Assuming that such change would be made, yes, it would have to be in the code generator. I'd like to bring up two points though:

1 - This would be a breaking change in the current design of the SDK, and we don't have plans for a new SDK version soon
2 - We are not convinced that this would make significant impact in reducing the Lambda cold-start, as JAR size is not the only factor to consider. You obviously read the recommendations we had in this [blog post](https://aws.amazon.com/blogs/developer/tuning-the-aws-java-sdk-2-x-to-reduce-startup-time/), but have you looked into using the SDK with GraalVM? You can read more details in this other post - https://aws.amazon.com/blogs/developer/graalvm-native-image-support-in-the-aws-sdk-for-java-2-x/
